aptos-move/move-examples/data_structures/sources/smart_table.move

```diff
@@ -0,0 +1,358 @@
+        bucket_size: u64,
+    }
+
+    /// Create an empty SmartTable with `initial_buckets` buckets.
+    public fun new<K: copy + drop + store, V: store>(): SmartTable<K, V> {
+        new_with_config<K, V>(0, 0, 0)
+    }
+
+    /// num_initial_buckets: The number of buckets on initialization. 0 means using default value.
+    /// split_load_threshold: The percent number which once reached, split will be triggered. 0 means using default value.
+    /// bucket_size: The target number of entries per bucket, though not enforced. 0 means not set and will dynamically
+    ///              assgined by the contract code.
+    public fun new_with_config<K: copy + drop + store, V: store>(num_initial_buckets: u64, split_load_threshold: u8, bucket_size: u64): SmartTable<K, V> {
+        assert!(split_load_threshold <= 100, error::invalid_argument(EINVALID_LOAD_THRESHOLD_PERCENT));
+        let buckets = table_with_length::new();
+        table_with_length::add(&mut buckets, 0, vector::empty());
+        let table = SmartTable {
+            buckets,
+            num_buckets: 1,
+            level: 0,
+            size: 0,
+            // The default split load threshold is 75%.
+            split_load_threshold: if (split_load_threshold == 0) {75} else {split_load_threshold},
+            bucket_size,
+        };
+        // The default number of initial buckets is 2.
+        if (num_initial_buckets == 0) {
+            num_initial_buckets = 2;
+        };
+        split(&mut table, num_initial_buckets- 1);
+        table
+    }
+
+    /// Destroy empty table.
+    /// Aborts if it's not empty.
+    public fun destroy_empty<K, V>(table: SmartTable<K, V>) {
+        assert!(table.size == 0, error::invalid_argument(ENOT_EMPTY));
+        let i = 0;
+        while (i < table.num_buckets) {
+            vector::destroy_empty(table_with_length::remove(&mut table.buckets, i));
+            i = i + 1;
+        };
+        let SmartTable {buckets, num_buckets: _, level: _, size: _, split_load_threshold:_, bucket_size: _} = table;
+        table_with_length::destroy_empty(buckets);
+    }
+
+    /// Destroy a table completely when V is dropable.
+    public fun destroy<K: drop, V: drop>(table: SmartTable<K, V>) {
+        let i = 0;
+        while (i < table.num_buckets) {
+            let j = 0;
+            let bucket = table_with_length::remove(&mut table.buckets, i);
+            let bucket_length = vector::length(&bucket);
+            while (j < bucket_length) {
+                vector::pop_back(&mut bucket);
+                j = j + 1;
+            };
+            vector::destroy_empty(bucket);
+            i = i + 1;
+        };
+        let SmartTable {buckets, num_buckets: _, level: _, size: _, split_load_threshold:_, bucket_size: _} = table;
+        table_with_length::destroy_empty(buckets);
+    }
+
+    /// Add (key, value) pair in the hash map, it may grow one bucket if current load factor exceeds the threshold.
```

Comment:
I think we should document the gas implication of a new bucket being created. May be we should also add a (getter) function that returns whether adding the next item would require creating a new bucket?

Reply:
Having more getter/view functions is always better :)